lib/azure/armrest/wrapper/base_wrapper.rb

```diff
@@ -0,0 +1,77 @@
+require 'delegate'
+require 'ostruct'
+
+module Azure
+  module ArmRest
+    # Base class for JSON wrapper classes. Each Manager class should have
+    # a corresponding class that wraps the JSON it collects, and each of
+    # them should subclass this base class.
+    class BaseWrapper < Delegator
+      # Access the json instance variable directly.
+      attr_accessor :json
+
+      # Constructs and returns a new JSON wrapper class. Pass in a plain
+      # JSON string and it will automatically give you accessor methods
+      # that make it behave like a typical Ruby object.
+      #
+      # Example:
+      #   class Person < Azure::ArmRest::BaseJsonWrapper; end
+      #
+      #   json_string = '{"firstname":"jeff", "lastname":"durand",
+      #     "address": { "street":"22 charlotte rd", "zipcode":"01013"}
+      #   }'
+      #
+      #   # Or whatever your subclass happens to be.
+      #   person = Person.new(json_string)
+      #
+      #   # The JSON properties are now available as methods.
+      #   person.firstname        # => 'jeff'
+      #   person.address.zipcode  # => '01013'
+      #
+      #   # Or you can get back the original JSON if necessary.
+      #   person.json # => Returns original JSON
+      #
+      def initialize(json)
+        @json = json
+        @ostruct = JSON.parse(json, object_class: OpenStruct)
+        __setobj__(@ostruct)
+      end
+
+      # Returns the original JSON string passed to the constructor.
+      def to_json
+        @json
+      end
+
+      # Explicitly convert the object to the original JSON string.
+      def to_s
+        @json
+      end
+
+      # Implicitly convert the object to the original JSON string.
+      def to_str
+        @json
+      end
+
+      protected
+
+      # Interface method required to make delegation work. Do
+      # not use this method directly.
+      def __getobj__
+        @ostruct
+      end
+
+      # A custom Delegator interface method that creates snake_case
+      # versions of the camelCase delegate methods.
+      def __setobj__(obj)
+        obj.methods(false).each{ |m|
+          if m.to_s[-1] != '=' # Must deal with nested ostruct's
+            res = obj.send(m)
+            __setobj__(res) if res.is_a?(OpenStruct)
+          end
+          snake = m.to_s.gsub(/(.)([A-Z])/,'\1_\2').downcase.to_sym
+          obj.instance_eval("alias #{snake} #{m}")
+        }
```

Comment:
I guess it really comes down to the usage of the methods that are being generated.  If we only call them once it turns out the overhead to dynamically define a method (which is what OpenStruct does) is not worth it.  In the ManageIQ code base, we just use method_missing and the overhead is removed.  If we are going to call the method like 100s of times however, then it's worth it to create the method.

See https://github.com/ManageIQ/manageiq/blob/master/lib/util/miq-hash_struct.rb for a class I had written that is like OpenStruct but uses method_missing instead of creating methods.  I remember when we implemented it, we got enormous performance increase, but that was also back on Ruby 1.8.7 :smile:   I had always intended on getting that back into Ruby proper somehow (like as an option to OpenStruct like `:generate_methods => false`) but never got around to it.

There may be an existing gem that does this kind of Struct-wrapping-without-generating-methods thing, but I haven't checked since.